File: 1394.find-lucky-integer-in-an-array.py
# @leet imports start
from string import *
from re import *
from datetime import *
from collections import *
from heapq import *
from bisect import *
from copy import *
from math import *
from random import *
from statistics import *
from itertools import *
from functools import *
from operator import *
from io import *
from sys import *
from json import *
from builtins import *
from typing import *
# @leet imports end


# @leet start
class Solution:
    def findLucky(self, arr: List[int]) -> int:
        freq = {}
        for num in arr:
            if num in freq:
                freq[num] += 1
            else:
                freq[num] = 1

        lucky_num = -1
        for num in freq.keys():
            if num == freq[num]:
                if num > lucky_num:
                    lucky_num = num
        return lucky_num


# @leet end

File: 1394.find-lucky-integer-in-an-array.js
// @leet start
/**
 * @param {number[]} arr
 * @return {number}
 */
var findLucky = function (arr) {
  const eachCount = new Map();
  const luckyNumbers = new Set();

  for (let i = 0; i < arr.length; i++) {
    const number = arr[i];
    let count;
    if (eachCount.has(number)) {
      count = eachCount.get(number);
      eachCount.set(number, count + 1);
      count += 1;
    } else {
      eachCount.set(number, 1);
      count = 1;
    }

    if (number == count) {
      luckyNumbers.add(number);
    }

    if (number != count && luckyNumbers.has(number)) {
      luckyNumbers.delete(number);
    }
  }

  return [...luckyNumbers.values()].length > 0
    ? Math.max(...luckyNumbers.values())
    : -1;
};
// @leet end

